File: client/src/main/java/net/treset/minecraftlauncher/util/string/UrlString.java
package net.treset.minecraftlauncher.util.string;

import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

public class UrlString extends FormatString {
    private final String original;
    private final boolean encode;

    public UrlString(String original, boolean encode) {
        this.original = original;
        this.encode = encode;
    }

    @Override
    public String get() throws FormatException {
        if(encode) {
            return URLEncoder.encode(original, StandardCharsets.UTF_8);
        } else {
            try {
                return URLDecoder.decode(original, StandardCharsets.UTF_8);
            } catch (IllegalArgumentException e) {
                throw new FormatException("Failed to decode URL string: " + original, e);
            }
        }
    }

    public String getOriginal() {
        return original;
    }

    public static UrlString encoded(String original) {
        return new UrlString(original, true);
    }

    public static UrlString decoded(String original) {
        return new UrlString(original, false);
    }
}

File: client/composeApp/src/commonMain/kotlin/dev/treset/treelauncher/backend/util/string/UrlString.kt
package dev.treset.treelauncher.backend.util.string

import java.net.URLDecoder
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

class UrlString(val original: String, private val operation: UrlOperation) : FormatString() {
    enum class UrlOperation {
        ENCODE,
        DECODE,
        NONE
    }

    @Throws(FormatException::class)
    override fun get(): String {
        return when (operation) {
            UrlOperation.ENCODE -> URLEncoder.encode(original, StandardCharsets.UTF_8)
            UrlOperation.DECODE -> try {
                    URLDecoder.decode(original, StandardCharsets.UTF_8)
                } catch (e: IllegalArgumentException) {
                    throw FormatException("Failed to decode URL string: $original", e)
                }
            else -> original
        }
    }

    companion object {
        fun of(original: String): UrlString {
            return UrlString(original, UrlOperation.NONE)
        }

        fun encoded(original: String): UrlString {
            return UrlString(original, UrlOperation.ENCODE)
        }

        fun decoded(original: String): UrlString {
            return UrlString(original, UrlOperation.DECODE)
        }
    }
}